### PR TITLE
Refactor kubernetes autodiscover to avoid skipping short-living pods

### LINKED ISSUE #22718: Kubernetes autodiscover doesn't discover short living jobs (and pods?)
Reported in https://github.com/elastic/beats/issues/11834#issuecomment-731024583. I could reproduce it in 7.10 with the reference documentation and a simple cronjob. Filebeat with autodiscover doesn't collect logs of short-living jobs.

What I could see is that short living containers don't generate any start event. Sometimes they generate a stop event.

With kubectl (`-w`) I could see that pods from short-living cronjobs don't generate an event with the running state:
```
hello-1606148340-spnpj   0/1     Pending     0          1s
hello-1606148340-spnpj   0/1     Pending     0          1s
hello-1606148340-spnpj   0/1     ContainerCreating   0          1s
hello-1606148340-spnpj   0/1     Completed           0          2s
```

With longer-living pods, this is the sequence of events seen, and logs are collected:
```
hello-1606148400-gm82x   0/1     Pending             0          0s
hello-1606148400-gm82x   0/1     Pending             0          0s
hello-1606148400-gm82x   0/1     ContainerCreating   0          0s
hello-1606148400-gm82x   1/1     Running             0          1s
hello-1606148400-gm82x   0/1     Completed           0          11s
```

I have also seen that logs from pods that print something and fail fast are not collected, events for these cases are like this:
```
echo   0/1     Pending   0          0s
echo   0/1     Pending   0          0s
echo   0/1     ContainerCreating   0          0s
echo   0/1     Completed           0          3s
echo   0/1     Completed           1          4s
echo   0/1     CrashLoopBackOff    1          5s
...
```
For these cases having the logs is important to help investigating what is happening.

If there are init containers, there can be cases where the logs for the init containers are not collected, in these cases event sequences like these ones are seen:
```
mytarget2                                            0/1     Init:0/1   0          6s
mytarget2                                            0/1     PodInitializing   0          15s
mytarget2                                            1/1     Running           0          19s
```

For Metricbeat it can be ok to don't start modules for short-living processes, but filebeat should collect logs of containers from the moment they start, it is important to investigate issues.

For confirmed bugs, please report:
- Version: 7.10.0 (also reported with 7.9.3)
- Discuss Forum URL: https://github.com/elastic/beats/issues/11834#issuecomment-731024583
- Steps to Reproduce: 
  * Start filebeat with the reference configuration and an autodiscover template:
    ```
        filebeat.autodiscover:
          providers:
            - type: kubernetes
              node: ${NODE_NAME}
              templates:
                - config:
                    - type: container
                      paths:
                        - /var/log/containers/*${data.kubernetes.container.id}.log
    ```
  * Run a cronjob with a short-living process, like this:
    ```
    apiVersion: batch/v1beta1
    kind: CronJob
    metadata:
      name: hello
    spec:
      schedule: "*/1 * * * *"
      failedJobsHistoryLimit: 10
      successfulJobsHistoryLimit: 20
      jobTemplate:
        spec:
          template:
            spec:
              containers:
              - name: hello
                image: busybox
                args:
                - /bin/sh
                - -c
                - date; echo Hello from the Kubernetes cluster
              restartPolicy: OnFailure
    ```

With debug logs for autodiscover this is seen for some jobs, some errors regarding the lack of `container.id`, and some stop events, but no start event:
```
2020-11-23T16:44:10.377Z	DEBUG	[autodiscover]	template/config.go:156	Configuration template cannot be resolved: field 'data.kubernetes.container.id' not available in event or environment accessing 'paths' (source:'/etc/filebeat.yml')
17:44:10.377
2020-11-23T16:44:10.377Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:236	Got a stop event: map[config:[] host:10.0.6.13 id:fedc9c0a-113c-414c-95e9-409b3e56ead8 kubernetes:{"annotations":{},"labels":{"controller-uid":"d0b5a9f3-cccc-41c1-90da-7bd4decbcf8c","job-name":"hello-1606149780"},"namespace":"cron","node":{"name":"gke-jsoriano-test-default-pool-a6f338c6-w0b6"},"pod":{"name":"hello-1606149780-bngj7","uid":"fedc9c0a-113c-414c-95e9-409b3e56ead8"}} meta:{"kubernetes":{"labels":{"controller-uid":"d0b5a9f3-cccc-41c1-90da-7bd4decbcf8c","job-name":"hello-1606149780"},"namespace":"cron","node":{"name":"gke-jsoriano-test-default-pool-a6f338c6-w0b6"},"pod":{"name":"hello-1606149780-bngj7","uid":"fedc9c0a-113c-414c-95e9-409b3e56ead8"}}} ports:{} provider:d8bb0011-c4ab-4e20-890c-e7a9ff56dfff stop:true]
17:44:10.377
2020-11-23T16:44:10.377Z	DEBUG	[autodiscover]	autodiscover/autodiscover.go:236	Got a stop event: map[config:[0xc000763e90] host:10.0.6.13 id:fedc9c0a-113c-414c-95e9-409b3e56ead8.hello kubernetes:{"annotations":{},"container":{"id":"054c10b6b0c8d8530735d3a92bbff5d76f4f76420e9c33319e5a0551be0fbf87","image":"busybox","name":"hello","runtime":"docker"},"labels":{"controller-uid":"d0b5a9f3-cccc-41c1-90da-7bd4decbcf8c","job-name":"hello-1606149780"},"namespace":"cron","node":{"name":"gke-jsoriano-test-default-pool-a6f338c6-w0b6"},"pod":{"name":"hello-1606149780-bngj7","uid":"fedc9c0a-113c-414c-95e9-409b3e56ead8"}} meta:{"container":{"id":"054c10b6b0c8d8530735d3a92bbff5d76f4f76420e9c33319e5a0551be0fbf87","image":{"name":"busybox"},"runtime":"docker"},"kubernetes":{"container":{"image":"busybox","name":"hello"},"labels":{"controller-uid":"d0b5a9f3-cccc-41c1-90da-7bd4decbcf8c","job-name":"hello-1606149780"},"namespace":"cron","node":{"name":"gke-jsoriano-test-default-pool-a6f338c6-w0b6"},"pod":{"name":"hello-1606149780-bngj7","uid":"fedc9c0a-113c-414c-95e9-409b3e56ead8"}}} port:0 provider:d8bb0011-c4ab-4e20-890c-e7a9ff56dfff stop:true]
```

### PR DESCRIPTION
**Pull request by jsoriano**
## What does this PR do?

Refactor logic in kubernetes autodiscover that decides when to generate events to try to address #22718.

Kubernetes autodiscover can generate events without network information now (without `host` or `port`/`ports`). This allows to generate events for pods that haven't started yet, or have succeeded/failed before generating a running event. These events still include the container id, so they can be used to collect logs. Still, no start event is generated if no pod ip and no container ids are available.

Some helpers have been added to obtain relevant information from pods and their containers.

Some additional small refactors are done to improve readability.

## Why is it important?

Current logic is checking similar things at the pod and container levels, try to simplify this logic focusing in containers only.

No matter what is the state of the pod, if there is a container running or trying to run, even if it is unhealthy, some configuration should be generated, so logs can be collected.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] ~~I have made corresponding changes to the documentation~~
- [x] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works. But more tests would be needed.
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Fix update so it definitely stops configurations, respecting the cleanup timeout.
- [x] Double-check that nothing breaks with events without `data.host`.
  This is logged at the debug level:
  ```
  2021-04-08T12:53:34.453+0200	DEBUG	[autodiscover]	template/config.go:156	Configuration template cannot be resolved: field 'data.host' not available in event or environment accessing 'hosts' (source:'/home/jaime/tmp/metricbeat-autodiscover-k8s.yml')
  ```
- [x] Actually check that the refactor solves the issues and doesn't break anything else.
- [x] Try to add more testing. A couple of cases added to current tests.
- [x] Check that nothing breaks with hints including `data.host` or `data.ports`.
  - [x] `data.ports.*` doesn't seem to be working, check if this is a regression.  Named ports were not added to container events. They are now.
  - [x] Multiple configurations generated for pods with multiple containers, expected? compare with master. False alarm.
- [x] Check autodiscover with heartbeat.

## Author's notes for the future

- When using autodiscover templates, if the condition matches pod conditions, container events will match too because they also include pod metadata. In pods with multiple containers this may lead to events with metadata of the incorrect container. This could be possibly avoided by matching per pod metadata only in the pod events. This is also happening in master, not solving in this PR. 
- `data.ports` is not included in container-level events, so it doesn't work when conditions for specific containers are used.

## How to test this PR locally

* Check that everything keeps running with some autodiscover happy cases.
* Check that logs are collected from a pod in CrashLoopBackOff (`kubectl run --image=redis redis -- echo foo`).

## Related issues

- Closes #22718.

## Use cases

Collect logs from containers in short-living or failing pods.

### CODE REVIEW

**Comment by hukaixuan** [comment]
Face the same issue, thanks for your great job! When will this refactor be done?

**Comment by jsoriano** [comment]
@hukaixuan thanks for your interest, I expect to dedicate some time to this in the following days, but I cannot give a fixed timeline, stay tunned to this PR :slightly_smiling_face:

**Comment by jsoriano** [comment]
> Hey @jsoriano !
> encountered the same issue as well, appreciate your work here!
> do you have a docker image of your changes I can use to test on my end?

Hi, this is still a draft, it still doesn't cover all cases, I actually know it doesn't work well with some basic cases.

File: libbeat/common/kubernetes/watcher.go [hunk 1]
```diff
@@ -97,16 +97,15 @@ type watcher struct {
...
-	informer, objType, err := NewInformer(client, resource, opts, indexers)
+	informer, _, err := NewInformer(client, resource, opts, indexers)
 	if err != nil {
 		return nil, err
 	}
 
 	store = informer.GetStore()
-	queue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), objType)
-	ctx, cancel := context.WithCancel(context.Background())
+	queue = workqueue.New()
```
Comment (jsoriano) [inline libbeat/common/kubernetes/watcher.go:108]: Not sure why a rate limited queue is needed here. Removing this for simplification, but happy to move the change to a different PR if preferred.

**Comment by jsoriano** [comment]
Opening for a first round of reviews, but I still see some weird things with `data.ports` and with pods with multiple containers. I have to compare with previous versions.

@ChrsMark it'd be great if you could take a look, thanks!

**Review by ChrsMark — COMMENTED** [review]
Thank you for improving all this codebase! I left some comments but it should be ok after addressing them.

One general thought is that I see that we have a lot of logic explained in comments here and there within the code. How delays work, how and why we emit stop/start events. These all look like a wider algorithm that we have to implement this feature, and we have special cases like `PodTerminated then do that` which can only be figured out is one look into the code and try to follow the flow.

I wonder if we could gather all this logic/algorithm into a more abstract document so as to have it documented centrally there. In this we will be able to run back to it when we have any kind of questions/possible bugs etc instead of interpreting the code ad-hoc. This might add an extra cost of maintaining this document but its up to us to decide to add this extra cost in order to have something useful for the future both for us but for upcoming people.
This is just a proposal since we are on it right now and can be added in another PR so as to avoid blocking this one.

**Comment by exekias** [comment]
@vjsamuel FYI

File: libbeat/autodiscover/providers/kubernetes/pod.go [hunk 2]
```diff
@@ -284,130 +262,123 @@ func getContainersInPod(pod *kubernetes.Pod) ([]kubernetes.Container, []kubernet
...
-		// Without this check there would be overlapping configurations with and without ports.
-		if len(c.Ports) == 0 {
-			// Set a zero port on the event to signify that the event is from a container
+		containerEvent := func(portName string, port int32) bus.Event {
 			event := bus.Event{
 				"provider":   p.uuid,
 				"id":         eventID,
 				flag:         true,
-				"host":       host,
-				"port":       0,
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go:384]: This port is still set to zero some lines below if [the container is running](https://github.com/elastic/beats/pull/24742/files#diff-ce319a56cef2bfe39d4095a22160b27ec8e769020ad2de71c31ae9f3e85e912bR357), and [there are no ports defined](https://github.com/elastic/beats/pull/24742/files#diff-ce319a56cef2bfe39d4095a22160b27ec8e769020ad2de71c31ae9f3e85e912bR373).

Good catch in any case, I overlooked heartbeat autodiscover, I'll check that it continues working.

[the remark this answers is not in the document]
Comment (ChrsMark) [reply libbeat/autodiscover/providers/kubernetes/pod.go:384]: 👍🏼

**Comment by jsoriano** [comment]
Going back to draft, I am reorganizing the code so the long `emit` function is splitted in easier to document functions.

File: libbeat/autodiscover/providers/kubernetes/pod.go [hunk 3]
```diff
@@ -114,59 +114,35 @@ func NewPodEventer(uuid uuid.UUID, cfg *common.Config, client k8s.Interface, pub
...
 }
 
-// OnAdd ensures processing of pod objects that are newly added
+// OnAdd ensures processing of pod objects that are newly added.
 func (p *pod) OnAdd(obj interface{}) {
 	p.logger.Debugf("Watcher Pod add: %+v", obj)
+	// Stop configs in case a re-sync generates an add event for a completed pod.
+	// Completed pods can happen with cronjobs for example, they are finished but not deleted
+	// by default.
+	// For this case this will always trigger a delayed stop, that will be executed after the
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go]: I have reverted changes here, they are not relevant for what this PR attempts to solve.

File: libbeat/autodiscover/providers/kubernetes/pod.go [hunk 4]
```diff
@@ -284,130 +262,123 @@ func getContainersInPod(pod *kubernetes.Pod) ([]kubernetes.Container, []kubernet
 	}
 	statuses = append(statuses, pod.Status.EphemeralContainerStatuses...)
 
-	return containers, statuses
+	containersInPod := make([]*containerInPod, len(containers))
+	for i, c := range containers {
+		cp := containerInPod{spec: c}
+		containersInPod[i] = &cp
+
+		// This nested loop can be a problem in pods with hundreds of containers, does it exist?
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go]: Refactored to use a map again to avoid the worst case scenario, and for loops to avoid generating so many temporary objects, as proposed in https://github.com/elastic/beats/pull/24742#discussion_r611388219.

Sorting the arrays is actually a good suggestion but complicates things a little bit, and can be also too much for the common use case.

We can revisit this function in a follow up if we see that needs to be improved.

File: libbeat/common/kubernetes/watcher.go [hunk 5]
```diff
@@ -216,53 +216,44 @@ func (w *watcher) enqueue(obj interface{}, state string) {
...
-			w.handler.OnAdd(o)
-		case update:
-			w.handler.OnUpdate(o)
-		case delete:
-			w.handler.OnDelete(o)
-		}
+	var entry *item
+	var ok bool
+	if entry, ok = obj.(*item); !ok {
+		// w.queue.Forget(obj)
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/common/kubernetes/watcher.go]: More or less :slightly_smiling_face: This would be needed with the rate limiting queue, I left it so I don't forget if we go back to use it. Removing it.

File: libbeat/autodiscover/providers/kubernetes/pod.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go:384]: Tested with heartbeat and it seems to be working. I found an issue, but also affects master, opened a different PR for this: https://github.com/elastic/beats/pull/25073

**Comment by jsoriano** [comment]
Opening again for review. Changes since previous reviews:
* `emit` has been further refactored so the main logic is hopefully more clear. Events generation has been moved out to specific functions to improve readability. Comments have been added with the intention to explain what each one of these functions do (not how they do it).
* `getContainersInPod` refactored to avoid worst-case scenario of problems with pods with many containers, and to avoid the need of a temporary array with all the containers.
* Change in `OnAdd` has been reverted, we can consider handling this corner case later if needed. Not directly related to the issue this PR intends to fix. At the moment this will only have the risk of leaving inactive harvesters running till the pods are definitely deleted.

> I wonder if we could gather all this logic/algorithm into a more abstract document so as to have it documented centrally there. In this we will be able to run back to it when we have any kind of questions/possible bugs etc instead of interpreting the code ad-hoc.  This might add an extra cost of maintaining this document but its up to us to decide to add this extra cost in order to have something useful for the future both for us and for the upcoming people.
> This is just a proposal since we are on it right now and can be added in another PR so as to avoid blocking this one.

@ChrsMark let me know if current version of this logic is easier to understand. I would like to avoid having to document this because then we have two problems: complex logic, and the need to keep documentation updated to this logic :slightly_smiling_face:

File: libbeat/autodiscover/providers/kubernetes/pod.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChrsMark) [reply libbeat/autodiscover/providers/kubernetes/pod.go]: 👍🏼

File: libbeat/autodiscover/providers/kubernetes/pod.go [hunk 6]
```diff
@@ -257,194 +224,276 @@ func (p *pod) Stop() {
...
 				"kubernetes": meta,
+				"container":  cmeta,
 			},
 		}
-		p.publish([]bus.Event{event})
+		// Include network information only if the container is running,
+		// so templates that need network don't generate a config.
+		if c.status.State.Running != nil {
+			if port.Name != "" && port.ContainerPort != 0 {
+				event["ports"] = common.MapStr{port.Name: port.ContainerPort}
```
Comment (ChrsMark) [inline libbeat/autodiscover/providers/kubernetes/pod.go:373]: 🤔 Will this just keep only the last "accessed" port of the loop because we create a new  `MapStr` each time?

**Review by ChrsMark — COMMENTED** [review]
Overall looks good so far 🍻 :
1. spotted one tricky line that might be a bug (maybe not :)). I might need to have a second pass with a clearer mind for hidden bug hunting (I believe there are none though)
2. @jroriano do you think you can improve testing notes a bit? I guess you have some specific scenarios you used to test this one. Would be nice to have them logged in this PR both for manual testing but also for future reference maybe.

File: libbeat/autodiscover/providers/kubernetes/pod.go (hunk 6 above)
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go:373]: Yes, I have changed the behaviour here, I have moved `data.ports` from the pod event to each container/port event. Let me know if you think this can be a problem.

Before, `data.ports` was included only in the pod event, so when configurations matched they didn't include the container metadata, only the pod metadata.

Code in this PR doesn't set the whole collection of `data.ports` in the pod event, it sets a single `data.ports.<portname>` field in to each container/port event, so when configurations match they include also the container metadata. By doing this each event with a named port is going to contain a single port.

This also fixes something I found misleading during my tests. As `data.ports` was included only in pod events, templates with conditions matching container fields, like the following one, didn't work. They work now with the changes here.
```
      templates:
        - condition:
            contains.kubernetes.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.ports.redis}"
```

**Comment by jsoriano** [comment]
>   2\. @jroriano do you think you can improve testing notes a bit? I guess you have some specific scenarios you used to test this one. Would be nice to have them logged in this PR both for manual testing but also for future reference maybe.

@ChrsMark I have some quite simple kubernetes manifests, that represent some cases in a list in a private manual test case. I would like to translate this at some moment to system or [e2e](https://github.com/elastic/e2e-testing) tests.

File: libbeat/autodiscover/providers/kubernetes/pod.go (hunk 6 above)
Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go:373]: You can see more clearly how this is different now in these test cases: https://github.com/elastic/beats/pull/24742/files#diff-9f24ba274864b933a726842c2b9fb8ee487a13545579938f9e373d0463611142R557

Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go:373]: Ok, what doesn't work now is this case. I will revert this change and do some more tests.
```
      templates:
        - condition:
            contains.kubernetes.pod.name: redis
          config:
            - module: redis
              hosts: "${data.host}:${data.ports.redis}"
            - module: nginx
              hosts: "${data.host}:${data.ports.http}"
```

We need more testing to cover all these cases :persevere:

Comment (ChrsMark) [reply libbeat/autodiscover/providers/kubernetes/pod.go:373]: Ah I see, it makes sense!

Comment (jsoriano) [reply libbeat/autodiscover/providers/kubernetes/pod.go]: Reverted in 330182725ac779372175f3ec897451935bbe6081, I will follow with this in a separate PR.

**Review by ChrsMark — APPROVED** [review]
lgtm, thank you!

File: libbeat/autodiscover/providers/kubernetes/pod.go [hunk 7]
```diff
@@ -257,194 +224,285 @@ func (p *pod) Stop() {
...
+		event := p.podEvent(flag, pod, portsMap, anyContainerRunning, annotations, namespaceAnnotations)
+		// Ensure that the pod level event is published first to avoid
+		// pod metadata overriding a valid container metadata.
+		eventList = append([][]bus.Event{{event}}, eventList...)
 	}
-	statuses = append(statuses, pod.Status.EphemeralContainerStatuses...)
 
-	return containers, statuses
+	delay := (flag == "stop" && podTerminated(pod, containers))
+	p.publishAll(eventList, delay)
```
Comment (exekias) [inline libbeat/autodiscover/providers/kubernetes/pod.go:311]: does it make sense to publish a `stop` event for not terminated pods?

Comment (exekias) [reply libbeat/autodiscover/providers/kubernetes/pod.go:311]: nevermind, this would be an update, got it

**Review by exekias — APPROVED** [review]
LGTM! let's make sure we test this thoroughly before releasing 😇

**Comment by jsoriano** [comment]
@ChrsMark @exekias Thanks for the reviews!

I did some additional tests and I found that the issue was not actually solved for hints-based autodiscover in filebeat, c9a3515 was missing. Could any of you please a quick review of this latest change?

**Review by ChrsMark — APPROVED** [review]
